mapo/envs/__init__.py

```diff
@@ -1,2 +1,107 @@
 # pylint: disable=missing-docstring
-from mapo.envs.navigation import NavigationEnv
+import abc
+import gym
+import tensorflow as tf
+
+# from mapo.envs.navigation import NavigationEnv
+
+
+class MAPOTFCustomEnv(gym.Env):
+    """MAPOCustomEnv defines an API on top of gym.Env in order
+    to initialize the necessary TensorFlow boilerplate in
+     _transition and _reward functions.
+
+    Args:
+        state_shape(Sequence[int]): The state shape.
+        action_shape(Sequence[int]): The action shape.
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, state_shape, action_shape):
+        self.state_shape = state_shape
+        self.action_shape = action_shape
+
+        self._state = None
+        self._timestep = None
+
+        self._graph = tf.Graph()
+        self._build_ops()
+        self._sess = tf.Session(graph=self._graph)
+
+    def _build_ops(self):
+
+        with self._graph.as_default():  # pylint: disable=not-context-manager
+
+            self._state_placeholder = tf.compat.v1.placeholder(
+                tf.float32, shape=self.state_shape
+            )
+            self._action_placeholder = tf.compat.v1.placeholder(
+                tf.float32, shape=self.action_shape
+            )
+            self._next_state_placeholder = tf.compat.v1.placeholder(
+                tf.float32, shape=self.state_shape
+            )
+
+            self._transition_tensor = self._transition_fn(
+                self._state_placeholder, self._action_placeholder
+            )
+            self._reward_tensor = self._reward_fn(
+                self._state_placeholder,
+                self._action_placeholder,
+                self._next_state_placeholder,
+            )
+
+    def step(self, action):
+        next_state = self._transition(self._state, action)
+        reward = self._reward(self._state, action, next_state)
+        done = self._terminal()
+        info = self._info()
+        self._state = next_state
+        self._timestep += 1
+        return self.obs, reward, done, info
+
+    def reset(self):
+        self._state = self.start_state
+        self._timestep = 0
+        return self.obs
+
+    def close(self):
+        self._sess.close()
+
+    @abc.abstractproperty
+    def start_state(self):
+        raise NotImplementedError
+
+    @abc.abstractproperty
+    def obs(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _transition_fn(self, state, action):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _reward_fn(self, state, action, next_state):
```

Comment:
What would be the best approach to use this function when we have batched `next_states` but just one `state` and `action`? This situation occurs when we do the one-step PGA loss and sample `config["branching_factor"]` next states. 

I suppose we shouldn't worry about it if the operation defined here broadcasts the `state` and `action` tensors to match the dimension of the `next_state` tensor.